### PR TITLE
Attempt to enable push to SSE encrypted s3 buckets 

### PR DESCRIPTION
**Pull request by aaelony**
Attempt to enable push to SSE encrypted s3 buckets (and not break for those s3 buckets that do not use SSE).

First PR from me ever so please review with a grain of salt! ;)
-A

### CODE REVIEW

File: src/onyx/plugin/s3_utils.clj [hunk 1]
```diff
@@ -26,13 +26,20 @@
...
-        metadata (doto (ObjectMetadata.)
-                   (.setContentLength size)
-                   ; (.setContentType contentType)
-                   (.setContentMD5 md5))
+        metadata (if (= encryption :aes256)
+                   (doto (ObjectMetadata.)
+                     (.setContentLength size)
+                     ;; (.setContentType contentType)
+                     (.setContentMD5 md5))
+                   (doto (ObjectMetadata.)
```
Comment (mariusz-jachimowicz-83) [inline src/onyx/plugin/s3_utils.clj]: Because of if statement this doto should be first, right ?

Comment (lbradstreet) [inline src/onyx/plugin/s3_utils.clj]: Yes, I believe it should be the other way around, which may the result of the CI failure.

**Comment by lbradstreet** [comment]
Please document the feature in https://github.com/onyx-platform/onyx-amazon-s3#output-task (default to `:none` in the sample) and https://github.com/onyx-platform/onyx-amazon-s3#attributes (describe options `:none` and `:aes256`. Thanks!

File: src/onyx/plugin/s3_utils.clj (hunk 1 above)
Comment (aaelony) [inline src/onyx/plugin/s3_utils.clj]: Sounds good.  I don't do a lot of java interop and rarely work with `doto`, defer to your guidance.

**Comment by aaelony** [comment]
Updated docs here: https://github.com/aaelony/onyx-amazon-s3/blob/aa-attempt-to-enable-optional-bucket-sse/README.md

File: src/onyx/plugin/s3_utils.clj (hunk 1 above)
Comment (MichaelDrogalis) [inline src/onyx/plugin/s3_utils.clj]: Would prefer to see the `if` be moved into a `case` to support multiple encryption schemes. Also drop out the commented code lines.

File: src/onyx/tasks/s3.clj [hunk 2]
```diff
@@ -28,7 +29,9 @@
   ([task-name :- s/Keyword
     bucket :- s/Str
     serializer-fn :- os/NamespacedKeyword
-    task-opts :- {s/Any s/Any}]
+    task-opts :- {s/Any s/Any}
+    encryption :- {(s/optional-key :aes256) s/Keyword}]
```
Comment (MichaelDrogalis) [inline src/onyx/tasks/s3.clj]: `:aes256` should be inferred from a list of allowed encryption schemes, shared with the conditional above.

**Comment by aaelony** [comment]
How would that be accomplished?

**Comment by aaelony** [comment]
with a case statement:

(defn upload [^TransferManager transfer-manager ^String bucket ^String
key ^clojure.lang.Keyword encryption ^bytes serialized
^ProgressListener progress-listener]
  (let [size (alength serialized)
        md5 (String. (Base64/encodeBase64 (DigestUtils/md5 serialized)))
        metadata (case encryption
                   :aes256 (doto (ObjectMetadata.)
                             (.setContentLength size)
                             (.setSSEAlgorithm "AES256")
                             (.setContentMD5 md5))
                   (doto (ObjectMetadata.)
                     (.setContentLength size)
                     (.setContentMD5 md5)))

        upload ^Upload (.upload transfer-manager
                                bucket
                                key
                                (ByteArrayInputStream. serialized)
                                metadata)]
    (.addProgressListener upload progress-listener)))

**Comment by lbradstreet** [comment]
I'd like to get this closed soon. Anyone around to help finish the job?

**Pull request merged**